### ISSUE TITLE
Include Plugins in Cocoapods checkout

### ISSUE THREAD

**Issue by Lutzifer** [grpc/grpc-swift#802]
Enhancement Proposal:

How about including precompiled versions of the plugins in the cocoapods checkout similar to how google does it with the crashlytics support tools 'run' and 'upload-symbols'? (see   https://github.com/firebase/firebase-ios-sdk/blob/master/FirebaseCrashlytics.podspec)

Compiling the protofiles could then be something like

```
for PROTO_FILE in $BASE_PATH/ProtobufDefinitions/*.proto; do
  protoc $PROTO_FILE \
		--proto_path=$BASE_PATH/ProtobufDefinitions \
		--plugin=./Pods/gRPC-Swift/bin/protoc-gen-swift \
		--swift_opt=Visibility=Public \
		--swift_out=$BASE_PATH/ProtobufGeneratedCode/
  protoc $PROTO_FILE \
  		--proto_path=$BASE_PATH/ProtobufDefinitions \
  		--plugin=./Pods/gRPC-Swift/bin/protoc-gen-grpc-swift \
  		--grpc-swift_opt=Visibility=Public \
      --grpc-swift_out=$BASE_PATH/GRPCGeneratedCode/
done;
```
(the most interesting part here is the use of 'Pods/gRPC-Swift/bin')
and the plugins and grpc-swift would always use versions that fit together.

**Comment by MrMage** [comment]
Given the amount of issues we have been encountering due to version mismatches between plugin and library, I like the idea. I am not an expert in how to best package that up, though, and I'm not sure if we have the bandwidth to tackle this right now 😅

But if you want to look into this, we would really appreciate it :-)

@glbrntt, what do you think?

**Comment by glbrntt** [comment]
I think this is a great idea; I also don't have bandwidth at the moment. @Lutzifer is this something you'd like to contribute?

**Comment by Lutzifer** [comment]
I will try to find the time to look into this.

Some Ideas:
- add a podspec for the plugins, which references gRPC-Swift.podspec as a dependency
- include the binary through this new plugin podspec, similar to how https://github.com/SwiftGen/SwiftGen/blob/master/SwiftGen.podspec does it
- missing part: how do we upload the compiled plugins and add them to a release? Code for this can be found at https://github.com/SwiftGen/SwiftGen/blob/c6477a6950d313c94c964c6c8ec960e3aeccb8f6/rakelib/release.rake#L132. Does this repo already have a script to make a release? In which order do you release what and how?

**Comment by MrMage** [comment]
> * missing part: how do we upload the compiled plugins and add them to a release? Code for this can be found at https://github.com/SwiftGen/SwiftGen/blob/c6477a6950d313c94c964c6c8ec960e3aeccb8f6/rakelib/release.rake#L132. Does this repo already have a script to make a release? In which order do you release what and how?

I think right now the process is just tagging a release on GitHub manually and running [the script that uploads a new spec to CocoaPods](https://github.com/grpc/grpc-swift/blob/master/scripts/build_podspecs.py); maybe building and uploading the plugins could be a part of that.

**Comment by Lutzifer** [comment]
I have begun working on this here: https://github.com/Lutzifer/grpc-swift/commit/755c410b2b143ce340650309151952da9041d18d

It is possible to test this by adding

```
pod "gRPC-Swift-Plugins", :podspec => 'https://raw.githubusercontent.com/Lutzifer/grpc-swift/feature/add_plugins_podspec/gRPC-Swift-Plugins.podspec'
```

in the podfile (in the main section, not in any target) and running pod update.

TODOs:
- [ ] Automate uploading built Plugins to releases
- [ ] Automate creating the podspec
- [ ] Make Plugins dependent on the main grpc podspec (not sure of this, yet)

**Comment by glbrntt** [comment]
@Lutzifer something that didn't occur to me when reviewing your PR for this: the binary will be unsigned so on macOS 10.15+ (I think) you'll have to manually trust the binary which isn't great.

We do plan on making the plugin available via Homebrew though (they don't allow pre-release tags, it seems). Would that be a better option altogether?

**Comment by Lutzifer** [comment]
Homebrew does not have a nicely useable system (or at least it did not when I last checked) for version management, so we would have decoupled versions of the pod and the plugins.

I invested a lot of time getting rid of Homebrew in our Development Environment as much as possible and migrated most of the tooling to mint to ensure correct versions.

**Comment by Lutzifer** [comment]
I also just checked on a coworkers machine, where we have no warning when using my branch of grpc-swift with the prebuilt plugins without any warning.

**Comment by glbrntt** [comment]
> I also just checked on a coworkers machine, where we have no warning when using my branch of grpc-swift with the prebuilt plugins without any warning.

Was it definitely an unsigned binary? When I checked between my two machines I _did_ get a warning and had to manually allow it via System Preferences. I've attached the same binaries. 

[protoc-grpc-swift-plugins-1.0.0-alpha.16.zip](https://github.com/grpc/grpc-swift/files/4918191/protoc-grpc-swift-plugins-1.0.0-alpha.16.zip)

**Comment by Lutzifer** [comment]
> > I also just checked on a coworkers machine, where we have no warning when using my branch of grpc-swift with the prebuilt plugins without any warning.
> 
> Was it definitely an unsigned binary? When I checked between my two machines I _did_ get a warning and had to manually allow it via System Preferences. I've attached the same binaries.
> 
> [protoc-grpc-swift-plugins-1.0.0-alpha.16.zip](https://github.com/grpc/grpc-swift/files/4918191/protoc-grpc-swift-plugins-1.0.0-alpha.16.zip)

I get this for my version (same result as for yours)

```
➜  gRPC-Swift-Plugins git:(master) ✗ codesign -dv --verbose=4 protoc-gen-grpc-swift 
protoc-gen-grpc-swift: code object is not signed at all
➜  gRPC-Swift-Plugins git:(master) ✗ codesign -dv --verbose=4 protoc-gen-swift                    [20/07/14|12:42PM]
protoc-gen-swift: code object is not signed at all
```

**Comment by glbrntt** [comment]
Huh, okay. Well I guess it doesn't matter too much why you didn't see this and I did. As long as you're okay with potentially having to manually approve the binary in the future then we should be okay to do this.

**Comment by glbrntt** [comment]
This has been resolved by #869 and #872
